Added flake8 plugins and some code fixes

mensinda opened the pull request:
This PR adds some flake8-plugins to sider to catch stuff like #5307 and raw `except:`. I am unsure about the `B007` warnings of [flake8-bugbear](https://pypi.org/project/flake8-bugbear/) but desided to keep them.

dcbaker commented on mesonbuild/dependencies/base.py:
```diff
@@ -924,9 +924,11 @@ def __repr__(self):
 
 class CMakeTarget:
-    def __init__(self, name, type, properies = {}):
+    def __init__(self, name, target_type, properies = None):
```
no space around the `=` in signatures unless they're type annotated:

```python
def foo(optional=None): ...
def bar(optional: Optional[List[str]] = None) -> None: ...
```

dcbaker commented on mesonbuild/interpreterbase.py:
```diff
@@ -144,7 +144,7 @@ def wrapped(*wrapped_args, **wrapped_kwargs):
 
 def noArgsFlattening(f):
-    setattr(f, 'no-args-flattening', True)
+    setattr(f, 'no-args-flattening', True) # noqa: B010
```
usually python style is to have two spaces between the last code line and the `#` in an inline comment.

mensinda replied:
Sure, but this is a magic comment that disables the `B010` and it has to be on the same line as the function.

Also, as far as I know, python doesn't have inline comments. I am not sure what you want me to change? Do you want me to add another ` ` in front of the `#`?

mensinda replied:
Nevermind, I found https://www.python.org/dev/peps/pep-0008/#inline-comments. I thought you were talking about C style comments.

dcbaker commented on mesonbuild/mlog.py:
```diff
@@ -183,7 +184,7 @@ def _log_error(severity, *args, **kwargs):
         args = (red('DEPRECATION:'),) + args
     else:
-        assert False, 'Invalid severity ' + severity
+        raise AssertionError('Invalid severity ' + severity)
```
I think this should be a `MesonException` or `TypeError`

mensinda replied:
Done

mensinda commented:
I have no idea why the vs 2017 and 2019 tests are failing. This PR should not have changed anything semantically...

dcbaker commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -455,7 +455,7 @@ def generate_target(self, target, outfile):
         # same time, also deal with generated sources that need to be compiled.
         generated_source_files = []
-        for rel_src, gensrc in generated_sources.items():
+        for rel_src, _ in generated_sources.items():
```
actually, better than this would be:
```python
for rel_src in generated_sources.keys():
```

mensinda replied:
Yeah, I also should have thought about that. Thanks.

QuLogic replied:
`for rel_src in generated_sources:` works as well.

mensinda replied:
Sure but using `.keys()` makes it obvious that we are iterating over a dict and not a list.

dcbaker commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -520,7 +520,7 @@ def generate_target(self, target, outfile):
 
         # Generate compile targets for all the pre-existing sources for this target
-        for f, src in target_sources.items():
+        for _, src in target_sources.items():
```
```python
for src in target_sources.values():
```

dcbaker commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -830,7 +830,7 @@ def generate_jar_target(self, target, outfile):
         generated_sources = self.get_target_generated_sources(target)
         gen_src_list = []
-        for rel_src, gensrc in generated_sources.items():
+        for rel_src, _ in generated_sources.items():
```
use .keys() here as well.

dcbaker commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -959,7 +959,7 @@ def determine_single_java_compile_args(self, target, compiler):
         deps = [os.path.join(self.get_target_dir(l), l.get_filename()) for l in target.link_targets]
         generated_sources = self.get_target_generated_sources(target)
-        for rel_src, gensrc in generated_sources.items():
+        for rel_src, _ in generated_sources.items():
```
.keys()

dcbaker commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -1148,7 +1148,7 @@ def generate_vala_compile(self, target, outfile):
                     target.install_dir[3] = os.path.join(self.environment.get_datadir(), 'gir-1.0')
         # Detect gresources and add --gresources arguments for each
-        for (gres, gensrc) in other_src[1].items():
+        for (_, gensrc) in other_src[1].items():
```
.values()

dcbaker commented on mesonbuild/backend/vs2010backend.py:
```diff
@@ -457,7 +457,7 @@ def add_project_reference(self, root, include, projid, link_outputs=False):
     def add_target_deps(self, root, target):
         target_dict = {target.get_id(): target}
-        for name, dep in self.get_target_deps(target_dict).items():
+        for _, dep in self.get_target_deps(target_dict).items():
```
.values()

dcbaker commented on mesonbuild/backend/xcodebackend.py:
```diff
@@ -345,7 +345,7 @@ def generate_pbx_file_reference(self):
 
     def generate_pbx_frameworks_buildphase(self):
-        for tname, t in self.build.targets.items():
+        for _, t in self.build.targets.items():
```
.values()

dcbaker commented on mesonbuild/backend/xcodebackend.py:
```diff
@@ -587,7 +587,7 @@ def generate_pbx_shell_build_phase(self, test_data):
     def generate_pbx_sources_build_phase(self):
         self.ofile.write('\n/* Begin PBXSourcesBuildPhase section */\n')
-        for name, phase_id in self.source_phase.items():
+        for name, _ in self.source_phase.items():
```
.keys()

dcbaker commented on mesonbuild/interpreter.py:
```diff
@@ -2193,7 +2193,7 @@ def get_variables(self):
         if self.build.environment.is_cross_build():
             props = self.build.environment.properties.host
-            for l, c in self.build.cross_compilers.items():
+            for l, _ in self.build.cross_compilers.items():
```
.keys()

ignatenkobrain commented on mesonbuild/mlog.py:
```diff
@@ -18,6 +18,7 @@
 import time
 import platform
 from contextlib import contextmanager
+from typing import Optional
```
__[Flake8]__ [F401] 'typing.Optional' imported but unused ([link](https://sider.review/gh/repos/19784232/pulls/5311#issue-4008812))

<sub>posted by Sider</sub>

mensinda commented:
OK, now this is weird... Why does the reverted change in 7d27ee2 cause the vs builds to fail? This is just a logging function! I honestly don't know what is going wrong here and can't really debug it because the [builds](https://dev.azure.com/jussi0947/jussi/_build/results?buildId=2383&view=logs&jobId=e635d226-97b8-55eb-2970-00a63a10ba8b&taskId=8a9b4f51-4992-5567-2aad-c06c4ac95ed0&lineStart=517&lineEnd=518&colStart=1&colEnd=1) fail during compilation and I have no knowlege of the VS backend(s).

It would be nice if someone else could have a look why this fails. Should I open a sepperate issue?

jon-turney commented:
> Why does the reverted change in [7d27ee2](https://github.com/mesonbuild/meson/commit/7d27ee20e0854debe6474c1394b1f589a71add3f) cause the vs builds to fail?

I believe that the presence of "ERROR" in the output causes msbuild to assume the command failed, irrespective of it's exit status.  (See also the discussion starting at https://github.com/mesonbuild/meson/issues/3095#issuecomment-412282630)

mensinda commented:
Rebased and fixed the VS errors by renaming `ERROR` to `Exception`. I have a deep admiration of the genius Microsoft programmers who implemented this excellent feature, without clearly documenting what is going on. Thanks @jon-turney for pointing this out. It probably would have taken me weeks of debugging to figure this one out.

jon-turney commented:
>  fixed the VS errors by renaming `ERROR` to `Exception`.

Unfortunately, I'm not sure this is a good idea, as a file:line reference followed by 'ERROR` has a special significance to some IDEs, making it a link to the source.

Considering this problem in the next larger context, since you haven't changed what `mlog.exception()` does, I guess that perhaps an exception is now being logged that was being caught by a blind except before?

mensinda commented:
@jon-turney I have already checked that this isn't caused by a previously uncaught exception. (However, I haven't checked yet if just moving the `red('ERROR:')` into the function body also triggers the failures in master). But, I can attibute the CI failures to the single change in 7d27ee20e0854debe6474c1394b1f589a71add3f.

I also haven't found out what specifically causes the error with the code moved into the function body. I am not sure how to fix this otherwise without changing the VS backend to ignore ERROR for meson.

mensinda commented:
I just found the real error: In [interpreter.py](https://github.com/mesonbuild/meson/blob/master/mesonbuild/interpreter.py#L2457) `mlog.exception` is called with `prefix=None`, which no longer disables the prefix with this change. I also checked all other functions to avoid running into a similar issue somewhere else.
